src/man/pkgdiff.1
'\" te
.\" Copyright (c) 2007, 2013, Oracle and/or its affiliates. All rights reserved.
.\" Copyright 2021 OmniOS Community Edition (OmniOSce) Association.
.Dd July 29, 2021
.Dt PKGDIFF 1
.Os
.Sh NAME
.Nm pkgdiff
.Nd compare package manifests
.Sh SYNOPSIS
.Nm
.Oo Fl i Ar attribute Oc Ns \&...
.Oo Fl o Ar attribute Oc Ns \&...
.Oo Fl t Ar action_name Ns Oo \&, Ns Ar action_name Oc Ns \&... Oc Ns \&...
.Oo Fl v Ar name Ns \&= Ns Ar value Oc Ns \&...
.Ar file1 file2
.Sh DESCRIPTION
.Nm
compares two package manifests and reports differences.
One of the manifests can be read from
.Dv stdin
by specifying
.Sy -
in place of the file name.
.Nm
sorts each manifest and action into a consistent order before comparison.
.Pp
Output is in the following form:
.Bl -tag -width Ds
.It Sy \&+ Ar complete_action
This action is in
.Ar file2
but not in
.Ar file1 .
.It Sy \&- Ar complete_action
This action is in
.Ar file1
but not in
.Ar file2 .
.It Sy \&- Ar attribute1 Ns = Ns Ar value1
This
.Ar attribute Ns \&, Ns Ar value
is in
.Ar file1
but not in
.Ar file2 .
.It Sy \&+ Ar attribute2 Ns = Ns Ar value2
This
.Ar attribute Ns \&, Ns Ar value
is in
.Ar file2
but not in
.Ar file1 .
.El
.Pp
Actions with different variants but the same type and key attribute value are
treated as separate actions for purposes of comparison.
Thus, actions that change attributes are shown in their complete form rather
than as attribute changes.
.Sh OPTIONS
The following options are supported:
.Bl -tag -width Ds
.It Fl \&-help
Display a usage message.
.It Fl i Ar attribute
Ignore
.Ar attribute
if present during comparisons.
Ignoring the
.Cm hash
attribute will also ignore any other hash attribute and, similarly, ignoring
.Cm size
will ignore all sizes.
Ignoring any attribute with this option will automatically also ignore the
.Cm timestamp
attribute.
This option cannot be used with the
.Fl o
option.
This option can be specified multiple times.
.It Fl o Ar attribute
Only report differences in
.Ar attribute .
This option cannot be used with the
.Fl i
option.
This option elides any action changes that do not affect
.Ar attribute
on an action.
This option can be specified multiple times.
.It Fl t Ar action_name Ns Oo \&, Ns Ar action_name Oc Ns \&...
Only compare the specified actions.
Multiple actions can be specified in a comma-separated list.
The value of
.Ar action_name
is one of the actions listed in "Actions" in the
.Xr pkg 7
man page.
For example, the following values could be specified for
.Ar action_name :
.Cm file ,
.Cm dir ,
.Cm driver ,
.Cm depend ,
.Cm set .
This option can be specified multiple times.
.It Fl v Ar name Ns = Ns Ar value
Only compute differences for this variant value.
For example, only compute differences for
.Cm arch=sparc .
This variant tag is removed for all actions before comparison.
Only one value can be specified per variant.
This option can be specified multiple times for different variants.
.El
.Sh EXIT STATUS
.Bl -tag -width Ds
.It Sy 0
No differences were found.
.It Sy 1
Differences were found.
.It Sy 2
Invalid command line options were specified.
.It Sy 3
An error occurred.
.It Sy 99
An unanticipated exception occurred.
.El
.Sh SEE ALSO
.Xr pkg 7
.Pp
.Lk https://github.com/omniosorg/pkg5


doc/history.txt

pkg
HISTORY

1. Summary

The intent of history is to enable pkg clients to maintain a record of all
image-modifying operations that they perform.

2. Discussion

2.1 Operations

    The following operations are currently recorded by the clients that are
    part of pkg(7):

        add-publisher
        image-create
        image-set-attributes
        image-update
        install
        purge-history
        rebuild-index
        refresh-publisher
        remove-publisher
        set-publisher
        set-preferred-publisher
        uninstall
        update-publisher

    These operations were chosen because they have the potential to alter the
    behavior of pkg(7) operations or because they modify an image.

3. History Entries

3.1 Information Recorded

    Each history entry contains a record of basic information about the client
    performing the operation, along with a mix of required and optional
    information about the operation performed.

    The following information about the client is always recorded:

        client_name     - The name of the client performing the operation
                          (e.g. pkg, packagemanager)

        client_version  - The value of pkg.VERSION at the time the client
                          performed the operation (e.g. 2e5300c4f0a4+)

        client_args     - The command-line arguments used when executing the
                          client (e.g. /usr/bin/pkg install foo)

    The following information about the operation performed is always recorded:

        operation_name  - The name of the operation performed (refer to 2.1)

        start_time      - When the operation started (e.g. 20080916T214726Z)

        end_time        - When the operation ended (e.g. 20080916T214800Z)

        userid          - The id of the user that performed the operation
                          (e.g. 0)

        username        - The username of the user that performed the operation
                          (e.g. root)

        result          - The outcome of the operation and the reason for it
                          (e.g. Failed, Transport) This maps to the pkg history
                          output columns of 'outcome' and 'reason'.

    The following information about the operation will be recorded if provided
    by the client or api:

        start_state     - Information about the operation requested before any
                          evaluation of that request is performed (e.g. an
                          image plan before evaluation)

        end_state       - Information about the operation requested after
                          evaluation of that request has been performed (e.g.
                          image plan after evaluation)

        errors          - Any errors that were encountered while performing the
                          operation (i.e. tracebacks, exceptions, etc.)

        be              - The name of the boot environment on which the
                          operation was performed

        be_uuid         - The uuid corresponding to the boot environment

        new_be          - The name of any new boot environment that was created
                          while performing the operation.

        new_be_uuid     - The uuid corresponding to the new boot environment

        snapshot        - The name of the snapshot that was taken as a result of
                          this operation.  If the operation completed
                          successfully and the snapshot was destroyed, this
                          this information is not stored.

3.2 Storage

    Each history entry is recorded in a XML file located in the metadata
    directory of an image (e.g. /var/pkg) in a directory named "history".
    Each XML file is named after the pattern %Y%m%dT%H%M%SZ-sequence.xml.
    Where sequence is a numeric value appended so that when multiple
    operations are performed on an image within the same second (e.g. -02,
    -03, etc.) entries are still written correctly.

3.3 File Format

    It should be noted that this format description is that of a private
    interface and is subject to change.

    History entries are recorded in a XML file with a fairly simplistic
    structure.  The format is as follows:

        The first line of each file is a standard xml header along with the
        encoding used to save the information.  This allows the pkg history
        command to correctly display this information if the client changes
        locales later.

        <?xml version="1.0" encoding="ascii"?>

        The second element of every history XML file is a root element used to
        contain the client and operation information.  Only one per file ever
        occurs.

        <history>

            The client element has a name and version attribute used to record
            client_name and client_version.  Only one per file ever occurs.

            <client name="pkg" version="2e5300c4f0a4+">

                The args element contains one or more "arg" (argument) elements
                that match each element of the system argument list used to
                execute the client.  It has no attributes.  Only one per file
                ever occurs.

                <args>

                    Each arg element contains a CDATA element to encapsulate
                    the raw information for the argument.  It has no attributes.

                    <arg>
                        <![CDATA[/usr/bin/pkg]]>
                    </arg>
                    <arg>
                        <![CDATA[-R]]>
                    </arg>
                    <arg>
                        <![CDATA[/tmp/test-image]]>
                    </arg>
                    <arg>
                        <![CDATA[install]]>
                    </arg>
                    <arg>
                        <![CDATA[SUNWvim]]>
                    </arg>
                </args>
            </client>


src/util/publish/pkgdiff.py
#!/usr/bin/python3 -Es
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

#
# Copyright (c) 2009, 2019, Oracle and/or its affiliates. All rights reserved.
#

from __future__ import print_function
import pkg.site_paths; pkg.site_paths.init()
import getopt
import gettext
import locale
import six
import sys
import traceback
import warnings
from functools import cmp_to_key

import pkg.actions
import pkg.variant as variant
import pkg.client.api_errors as apx
import pkg.manifest as manifest
import pkg.misc as misc
from pkg.misc import PipeError, CMP_UNSIGNED, CMP_ALL
from collections import defaultdict
from itertools import product
from pkg.client.pkgdefs import EXIT_OK, EXIT_OOPS, EXIT_BADOPT, EXIT_PARTIAL

def usage(errmsg="", exitcode=EXIT_BADOPT):
        """Emit a usage message and optionally prefix it with a more specific
        error message.  Causes program to exit."""

        if errmsg:
                print("pkgdiff: {0}".format(errmsg), file=sys.stderr)

        print(_("""\
Usage:
        pkgdiff [-i attribute]... [-o attribute]
            [-t action_name[,action_name]...]...
            [-v name=value]... (file1 | -) (file2 | -)"""))
        sys.exit(exitcode)

def error(text, exitcode=EXIT_PARTIAL):
        """Emit an error message prefixed by the command name """

        print("pkgdiff: {0}".format(text), file=sys.stderr)

        if exitcode != None:
                sys.exit(exitcode)

def main_func():

        ignoreattrs = []
        onlyattrs = []
        onlytypes = []
        varattrs = defaultdict(set)
        cmp_policy = CMP_ALL

        try:
                opts, pargs = getopt.getopt(sys.argv[1:], "i:o:t:uv:?", ["help"])
                for opt, arg in opts:
                        if opt == "-i":
                                ignoreattrs.append('timestamp')
                                ignoreattrs.append(arg)
                                if arg == 'hash':
                                        ignoreattrs.extend([
                                            'chash', 'pkg.content-hash',
                                            'elfhash', 'timestamp'])
                                if arg == 'size':
                                        ignoreattrs.extend([
                                            'pkg.csize', 'pkg.size'])
                        elif opt == "-o":
                                onlyattrs.append(arg)
                        elif opt == "-t":
                                onlytypes.extend(arg.split(","))
                        elif opt == "-u":
                                cmp_policy = CMP_UNSIGNED
                        elif opt == "-v":
                                args = arg.split("=")
                                if len(args) != 2:
                                        usage(_("variant option incorrect {0}").format(
                                            arg))
                                if not args[0].startswith("variant."):
                                        args[0] = "variant." + args[0]
                                varattrs[args[0]].add(args[1])
                        elif opt in ("--help", "-?"):
                                usage(exitcode=EXIT_OK)

        except getopt.GetoptError as e:
                usage(_("illegal global option -- {0}").format(e.opt))

        if len(pargs) != 2:
                usage(_("two manifest arguments are required"))

        if (pargs[0] == "-" and pargs[1] == "-"):
                usage(_("only one manifest argument can be stdin"))

        if ignoreattrs and onlyattrs:
                usage(_("-i and -o options may not be used at the same time."))

        for v in varattrs:
                if len(varattrs[v]) > 1:
                        usage(_("For any variant, only one value may be "
                            "specified."))
                varattrs[v] = varattrs[v].pop()

        ignoreattrs = set(ignoreattrs)
        onlyattrs = set(onlyattrs)
        onlytypes = set(onlytypes)

        utypes = set(
            t
            for t in onlytypes
            if t == "generic" or t not in pkg.actions.types
        )

        if utypes:
                usage(_("unknown action types: {0}".format(
                    apx.list_to_lang(list(utypes)))))

        manifest1 = manifest.Manifest()
        manifest2 = manifest.Manifest()
        try:
                # This assumes that both pargs are not '-'.
                for p, m in zip(pargs, (manifest1, manifest2)):
                        if p == "-":
                                m.set_content(content=sys.stdin.read())
                        else:
                                m.set_content(pathname=p)
        except (pkg.actions.ActionError, apx.InvalidPackageErrors) as e:
                error(_("Action error in file {p}: {e}").format(**locals()))
        except (EnvironmentError, apx.ApiException) as e:
                error(e)

        #
        # manifest filtering
        #

        # filter action type
        if onlytypes:
                for m in (manifest1, manifest2):
                        # Must pass complete list of actions to set_content, not
                        # a generator, to avoid clobbering manifest contents.
                        m.set_content(content=list(m.gen_actions_by_types(
                            onlytypes)))

        # filter variant
        v1 = manifest1.get_all_variants()
        v2 = manifest2.get_all_variants()
        for vname in varattrs:
                for _path, v, m in zip(pargs, (v1, v2), (manifest1, manifest2)):
                        if vname not in v:
                                continue
                        filt = varattrs[vname]
                        if filt not in v[vname]:
                                usage(_("Manifest {path} doesn't support "
                                    "variant {vname}={filt}".format(**locals())))
                        # remove the variant tag
                        def rip(a):
                                a.attrs.pop(vname, None)
                                return a
                        m.set_content([
                            rip(a)
                            for a in m.gen_actions(excludes=[
                            variant.Variants({vname: filt}).allow_action])
                        ])
                        m[vname] = filt

        if varattrs:
                # need to rebuild these if we're filtering variants
                v1 = manifest1.get_all_variants()
                v2 = manifest2.get_all_variants()

        # we need to be a little clever about variants, since
        # we can have multiple actions w/ the same key attributes
        # in each manifest in that case.  First, make sure any variants
        # of the same name have the same values defined.
        for k in set(v1.keys()) & set(v2.keys()):
                if v1[k] != v2[k]:
                        error(_("Manifests support different variants "
                            "{v1} {v2}").format(v1=v1, v2=v2))

        # Now, get a list of all possible variant values, including None
        # across all variants and both manifests
        v_values = dict()

        for v in v1:
                v1[v].add(None)
                for a in v1[v]:
                        v_values.setdefault(v, set()).add((v, a))

        for v in v2:
                v2[v].add(None)
                for a in v2[v]:
                        v_values.setdefault(v, set()).add((v, a))

        diffs = []

        for tup in product(*v_values.values()):
                # build excludes closure to examine only actions exactly
                # matching current variant values... this is needed to
                # avoid confusing manifest difference code w/ multiple
                # actions w/ same key attribute values or getting dups
                # in output
                def allow(a, publisher=None):
                        for k, v in tup:
                                if v is not None:
                                        if k not in a.attrs or a.attrs[k] != v:
                                                return False
                                elif k in a.attrs:
                                        return False
                        return True

                a, c, r = manifest2.difference(manifest1, [allow], [allow],
                    cmp_policy=cmp_policy)
                diffs += a
                diffs += c
                diffs += r

        # License action still causes spurious diffs... check again for now.
        real_diffs = [
            (a, b)
            for a, b in diffs
            if a is None or b is None or a.different(b, cmp_policy=cmp_policy)
        ]

        if not real_diffs:
                return 0

        # define some ordering functions so that output is easily readable
        # First, a human version of action comparison that works across
        # variants and action changes...
        def compare(a, b):
                # pull the relevant action out of the old value, new
                # value tuples
                a = a[0] if a[0] else a[1]
                b = b[0] if b[0] else b[1]

                if hasattr(a, "key_attr") and hasattr(b, "key_attr") and \
                    a.key_attr == b.key_attr:
                        res = misc.cmp(a.attrs[a.key_attr], b.attrs[b.key_attr])
                        if res != NotImplemented:
                                return res
                        # sort by variant
                        res = misc.cmp(sorted(list(a.get_variant_template())),
                            sorted(list(b.get_variant_template())))
                        if res != NotImplemented:
                                return res
                else:
                        res = misc.cmp(a.ordinality, b.ordinality)
                        if res != NotImplemented:
                                return res
                # Fall back to a simple string compare if we have
                # differing types (indicated by the above NotImplemented
                # return values.
                return misc.cmp(str(a), str(b))

        # sort and....
        diffs = sorted(diffs, key=cmp_to_key(compare))

        # handle list attributes
        def attrval(attrs, k, elide_iter=tuple()):
                def q(s):
                        if " " in s or s == "":
                                return '"{0}"'.format(s)
                        else:
                                return s

                v = attrs[k]
                if isinstance(v, list) or isinstance(v, set):
                        out = " ".join([
                            "{0}={1}".format(k, q(lmt))
                            for lmt in sorted(v)
                            if lmt not in elide_iter
                        ])
                elif " " in v or v == "":
                        out = k + "=\"" + v + "\""
                else:
                        out = k + "=" + v
                return out

        # figure out when to print diffs
        def conditional_print(s, a):
                if onlyattrs:
                        if not set(a.attrs.keys()) & onlyattrs:
                                return False
                elif ignoreattrs:
                        if not set(a.attrs.keys()) - ignoreattrs:
                                return False

                print("{0} {1}".format(s, a))
                return True

        different = False

        for old, new in diffs:
                if not new:
                        different |= conditional_print("-", old)
                elif not old:
                        different |= conditional_print("+", new)
                else:
                        s = []

                        if not onlyattrs:
                                if (hasattr(old, "hash") and
                                    "hash" not in ignoreattrs):
                                        if old.hash != new.hash:
                                                s.append("  - {0}".format(old.hash))
                                                s.append("  + {0}".format(new.hash))
                                attrdiffs = (set(new.differences(old)) -
                                    ignoreattrs)
                                attrsames = sorted( list(set(list(old.attrs.keys()) +
                                    list(new.attrs.keys())) -
                                    set(new.differences(old))))
                        else:
                                if hasattr(old, "hash") and "hash" in onlyattrs:
                                        if old.hash != new.hash:
                                                s.append("  - {0}".format(old.hash))
                                                s.append("  + {0}".format(new.hash))
                                attrdiffs = (set(new.differences(old)) &
                                    onlyattrs)
                                attrsames = sorted(list(set(list(old.attrs.keys()) +
                                    list(new.attrs.keys())) -
                                    set(new.differences(old))))

                        for a in sorted(attrdiffs):
                                if a in old.attrs and a in new.attrs and \
                                    isinstance(old.attrs[a], list) and \
                                    isinstance(new.attrs[a], list):
                                        elide_set = (set(old.attrs[a]) &
                                            set(new.attrs[a]))
                                else:
                                        elide_set = set()
                                if a in old.attrs:
                                        diff_str = attrval(old.attrs, a,
                                            elide_iter=elide_set)
                                        if diff_str:
                                                s.append("  - {0}".format(diff_str))
                                if a in new.attrs:
                                        diff_str = attrval(new.attrs, a,
                                            elide_iter=elide_set)
                                        if diff_str:
                                                s.append("  + {0}".format(diff_str))
                        # print out part of action that is the same
                        if s:
                                different = True
                                print("{0} {1} {2}".format(old.name,
                                    attrval(old.attrs, old.key_attr),
                                    " ".join(("{0}".format(attrval(old.attrs,v))
                                    for v in attrsames if v != old.key_attr))))

                                for l in s:
                                        print(l)

        return int(different)

if __name__ == "__main__":
        misc.setlocale(locale.LC_ALL, "", error)
        gettext.install("pkg", "/usr/share/locale",
            codeset=locale.getpreferredencoding())
        misc.set_fd_limits(printer=error)

        if six.PY3:
                # disable ResourceWarning: unclosed file
                warnings.filterwarnings("ignore", category=ResourceWarning)
        try:
                exit_code = main_func()
        except (PipeError, KeyboardInterrupt):
                exit_code = EXIT_OOPS
        except SystemExit as __e:
                exit_code = __e
        except Exception as __e:
                traceback.print_exc()
                error(misc.get_traceback_message(), exitcode=None)
                exit_code = 99

        sys.exit(exit_code)

# Vim hints
# vim:ts=8:sw=8:et:fdm=marker
